### ISSUE TITLE
False positives with unusual, non utf-8 characters

### ISSUE THREAD

**Issue by tyarosevich** [maxbachmann/rapidfuzz#138]
I encountered what seems to be a bug when trying to match a string against a list of strings which includes an element with non utf-8 characters. The following code should reproduce and illustrate the issue. Sorry if this isn't formatted correctly, I haven't submitted an issue on github before. I've included my local output below the raw code.

```
from rapidfuzz import fuzz, process

# Apply the problematic string as-is (origin is the Pubmed search API via the Entrezpy library
problem_string = '[Evaluation of the Filaria Detect™ IgG4 kit employing recombinant Wb123 antigen for diagnosis of lymphatic filariasis in BrazilEvaluación del estuche Detect™ para pruebas de anticuerpos antifiláricos IgG4, fabricado con el antígeno recombinante Wb123, para el diagnóstico de la filariasis linfática en Brasil].'
lst_choices = [problem_string]

# Example of a false positive. This string comes from the Web of Science API.
title = 'Biochemical characterization of Recombinase A from Wolbachia endosymbiont of filarial nematode Brugia malayi (wBmRecA).'
output = process.extract(title, lst_choices, scorer=fuzz.partial_ratio, score_cutoff=90)
print('Title compared to original string: \n {} \n'.format(output))

# Apply the problematic string with the trademark symbols removed.
problem_string = '[Evaluation of the Filaria Detect IgG4 kit employing recombinant Wb123 antigen for diagnosis of lymphatic filariasis in BrazilEvaluación del estuche Detect para pruebas de anticuerpos antifiláricos IgG4, fabricado con el antígeno recombinante Wb123, para el diagnóstico de la filariasis linfática en Brasil].'
lst_choices = [problem_string]
output = process.extract(title, lst_choices, scorer=fuzz.partial_ratio, score_cutoff=90)
print('Title compared to original string with TM symbols removed: \n {} \n'.format(output))

# Same code with the problematic string encoded to a byte string.
problem_string = '[Evaluation of the Filaria Detect™ IgG4 kit employing recombinant Wb123 antigen for diagnosis of lymphatic filariasis in BrazilEvaluación del estuche Detect™ para pruebas de anticuerpos antifiláricos IgG4, fabricado con el antígeno recombinante Wb123, para el diagnóstico de la filariasis linfática en Brasil].'
problem_string = problem_string.encode('utf-8')
lst_choices = [problem_string]

output = process.extract(title, lst_choices, scorer=fuzz.partial_ratio, score_cutoff=90)
print('Title compared to string encoded as byte string: \n {} \n'.format(output))

# Same comparison made for reference with the non-problematic string comppared to itself, with the title as a regular
# string, and the comparison title as a byte string.
lst_choices = [title.encode('utf-8')]
output = process.extract(title, lst_choices, scorer=fuzz.partial_ratio, score_cutoff=90)
print('Title compared to itself encoded as byte string: \n {} \n'.format(output))

```

Local output:
```
Title compared to original string: 
 [('[Evaluation of the Filaria Detect™ IgG4 kit employing recombinant Wb123 antigen for diagnosis of lymphatic filariasis in BrazilEvaluación del estuche Detect™ para pruebas de anticuerpos antifiláricos IgG4, fabricado con el antígeno recombinante Wb123, para el diagnóstico de la filariasis linfática en Brasil].', 100.0, 0)] 
Title compared to original string with TM symbols removed: 
 [] 
Title compared to string encoded as byte string: 
 [] 
Title compared to itself encoded as byte string: 
 [(b'Biochemical characterization of Recombinase A from Wolbachia endosymbiont of filarial nematode Brugia malayi (wBmRecA).', 100.0, 0)] 


```

**Comment by maxbachmann** [comment]
You are currently comparing  a string to a iterable of characters, which is probably not want you wanted to to. I assume, that you wanted to use:
```python
problem_string = ['Evaluation of the Filaria Detect™ IgG4 kit employing recombinant Wb123 antigen for diagnosis of lymphatic filariasis in BrazilEvaluación del estuche Detect™ para pruebas de anticuerpos antifiláricos IgG4, fabricado con el antígeno recombinante Wb123, para el diagnóstico de la filariasis linfática en Brasil.']
```
instead of:
```python
problem_string = '[Evaluation of the Filaria Detect™ IgG4 kit employing recombinant Wb123 antigen for diagnosis of lymphatic filariasis in BrazilEvaluación del estuche Detect™ para pruebas de anticuerpos antifiláricos IgG4, fabricado con el antígeno recombinante Wb123, para el diagnóstico de la filariasis linfática en Brasil].'
```

**Comment by maxbachmann** [comment]
Note that utf8 encoding is not supported. I compare strings by comparing their code points. Utf8 might use multiple code points to represent a single character, which would be interpreted as multiple characters by RapidFuzz.

**Comment by tyarosevich** [comment]
Hi Max, thank you very much for responding. I think my original code was slightly unclear - I was putting the problematic test title in a list in the function call itself, so I can see why you may have thought it was being passed directly at a glance. I have updated my example to reproduce the issue to more clearly demonstrate how the trademark symbol is causing an issue, and how converting to a byte string gets rid of it. I'm not an expert on these matters, but the problem was persistent across a dataset of thousands, all of which would produce false positives. Applying the fix I've shown above fixed the problem, so I figured I would update my example in case it isn't some sort of mistake on my part.

Incidentally, it's my understanding that all python strings use UTF-8 encoding by default, so I'm not sure that's the issue - again, I could just be ignorant, I'm just a data analyst that  does a lot of coding in python, not a dev.

P.S. thank you for this library, it's fantastic.

**Comment by maxbachmann** [comment]
> Incidentally, it's my understanding that all python strings use UTF-8 encoding by default, so I'm not sure that's the issue

Python internally stores strings using a flexible representation, which is described in https://www.python.org/dev/peps/pep-0393/

> I have updated my example to reproduce the issue to more clearly demonstrate how the trademark symbol is causing an issue, and how converting to a byte string gets rid of it

Thanks for the update. I can now reproduce this issue. I am not sure whats causing the bug yet (something in the get_matching_blocks implementation is broken). It breaks when one of the strings uses characters outside of the extended ascii range. A minimal example is:
```python
str1 = 'a'*65
str2 = 'a' + chr(256) + 'a'*63
fuzz.partial_ratio(str1, str2)
```
which returns a score of 100

**Comment by maxbachmann** [comment]
Thanks for reporting. This is fixed in [version 1.7.1](https://github.com/maxbachmann/RapidFuzz/releases/tag/v1.7.1).
